File: diarization.js
// Imports the Google Cloud client library
const speech = require('@google-cloud/speech').v1p1beta1;

// Creates a client
const client = new speech.SpeechClient();

/**
 * TODO(developer): Uncomment the following line before running the sample.
 */
// const uri = path to GCS audio file e.g. `gs:/bucket/audio.wav`;

const config = {
  encoding: 'LINEAR16',
  sampleRateHertz: 44100,
  languageCode: 'ja-JP',
  enableSpeakerDiarization: true,
  diarizationSpeakerCount: 3,
  model: 'command_and_search',
};

const audio = {
  uri: 'gs://0812_0/zoom_1_test.wav',
};

const request = {
  config: config,
  audio: audio,
};

async function main() {
  const [operation] = await client.longRunningRecognize(request);
  const [response] = await operation.promise();
  // const [response] = await client.recognize(request);
  const transcription = response.results
    .map(result => result.alternatives[0].transcript)
    .join('\n');
  console.log(`Transcription: ${transcription}`);
  console.log('Speaker Diarization:');
  const result = response.results[response.results.length - 1];
  const wordsInfo = result.alternatives[0].words;
  // Note: The transcript within each result is separate and sequential per result.
  // However, the words list within an alternative includes all the words
  // from all the results thus far. Thus, to get all the words with speaker
  // tags, you only have to take the words list from the last result:
  wordsInfo.forEach(a =>
    console.log(` word: ${a.word}, speakerTag: ${a.speakerTag}`)
  );
}

main()

File: Diarization.java
// Imports the Google Cloud client library
import com.google.cloud.speech.v1.RecognitionAudio;
import com.google.cloud.speech.v1.RecognitionConfig;
import com.google.cloud.speech.v1.RecognitionConfig.AudioEncoding;
import com.google.cloud.speech.v1.RecognizeResponse;
import com.google.cloud.speech.v1.SpeechClient;
import com.google.cloud.speech.v1.SpeechRecognitionAlternative;
import com.google.cloud.speech.v1.SpeechRecognitionResult;
import com.google.protobuf.ByteString;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

class Diarization {

  public static void main(String[] args) {
    // System.out.println("Hello, world.");
    // System.out.println(args[0]);
    System.out.println(args[0]);
    transcribeDiarizationGcs(args[0]);
  }

  /**
   * Transcribe a remote audio file using speaker diarization.
   *
   * @param gcsUri the path to an audio file.
   */
  public static void transcribeDiarizationGcs(String gcsUri) throws Exception {
    try (SpeechClient speechClient = SpeechClient.create()) {
      SpeakerDiarizationConfig speakerDiarizationConfig = SpeakerDiarizationConfig
        .newBuilder()
        .setEnableSpeakerDiarization(true)
        .setMinSpeakerCount(2)
        .setMaxSpeakerCount(3)
        .build();

      // Configure request to enable Speaker diarization
      RecognitionConfig config = RecognitionConfig
        .newBuilder()
        .setEncoding(AudioEncoding.LINEAR16)
        .setLanguageCode("ja-JP")
        .setSampleRateHertz(44100)
        .setDiarizationConfig(speakerDiarizationConfig)
        .build();

      // Set the remote path for the audio file
      RecognitionAudio audio = RecognitionAudio
        .newBuilder()
        .setUri(gcsUri)
        .build();

      // Use non-blocking call for getting file transcription
      OperationFuture<LongRunningRecognizeResponse, LongRunningRecognizeMetadata> response = speechClient.longRunningRecognizeAsync(
        config,
        audio
      );

      while (!response.isDone()) {
        System.out.println("Waiting for response...");
        Thread.sleep(10000);
      }

      // Speaker Tags are only included in the last result object, which has only one alternative.
      LongRunningRecognizeResponse longRunningRecognizeResponse = response.get();
      SpeechRecognitionAlternative alternative = longRunningRecognizeResponse
        .getResults(longRunningRecognizeResponse.getResultsCount() - 1)
        .getAlternatives(0);

      // The alternative is made up of WordInfo objects that contain the speaker_tag.
      WordInfo wordInfo = alternative.getWords(0);
      int currentSpeakerTag = wordInfo.getSpeakerTag();

      // For each word, get all the words associated with one speaker, once the speaker changes,
      // add a new line with the new speaker and their spoken words.
      StringBuilder speakerWords = new StringBuilder(
        String.format(
          "Speaker %d: %s",
          wordInfo.getSpeakerTag(),
          wordInfo.getWord()
        )
      );

      for (int i = 1; i < alternative.getWordsCount(); i++) {
        wordInfo = alternative.getWords(i);
        if (currentSpeakerTag == wordInfo.getSpeakerTag()) {
          speakerWords.append(" ");
          speakerWords.append(wordInfo.getWord());
        } else {
          speakerWords.append(
            String.format(
              "\nSpeaker %d: %s",
              wordInfo.getSpeakerTag(),
              wordInfo.getWord()
            )
          );
          currentSpeakerTag = wordInfo.getSpeakerTag();
        }
      }

      System.out.println(speakerWords.toString());
    }
  }
}

